streams/shared/src/main/scala/zio/stream/experimental/ZTransducer.scala

```diff
@@ -124,6 +128,223 @@ object ZTransducer {
+   * of type `O` until `max` elements have been folded.
+   *
+   * Like [[foldWeighted]], but with a constant cost function of 1.
+   */
+  def foldUntil[I, O](z: O, max: Long)(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    foldWeighted[I, O](z)(_ => 1, max)(f)
+
+  /**
+   * Creates a transducer that effectfully folds elements of type `I` into a structure
+   * of type `O` until `max` elements have been folded.
+   *
+   * Like [[foldWeightedM]], but with a constant cost function of 1.
+   */
+  def foldUntilM[R, E, I, O](z: O, max: Long)(f: (O, I) => ZIO[R, E, O]): ZTransducer[R, E, I, O] =
+    foldWeightedM[R, E, I, O](z)(_ => UIO.succeedNow(1), max)(f)
+
+  /**
+   * Creates a transducer that folds elements of type `I` into a structure
+   * of type `O`, until `max` worth of elements (determined by the `costFn`)
+   * have been folded.
+   *
+   * @note Elements that have an individual cost larger than `max` will
+   * cause the stream to hang. See [[foldWeightedDecompose]] for
+   * a variant that can handle these.
+   */
+  def foldWeighted[I, O](z: O)(costFn: I => Long, max: Long)(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    foldWeightedDecompose[I, O](z)(costFn, max, Chunk.single(_))(f)
+
+  /**
+   * Creates a transducer that folds elements of type `I` into a structure
+   * of type `O`, until `max` worth of elements (determined by the `costFn`)
+   * have been folded.
+   *
+   * The `decompose` function will be used for decomposing elements that
+   * cause an `O` aggregate to cross `max` into smaller elements. For
+   * example:
+   * {{{
+   * Stream(1, 5, 1)
+   *  .aggregate(
+   *    ZTransducer
+   *      .foldWeightedDecompose(List[Int]())((i: Int) => i.toLong, 4,
+   *        (i: Int) => Chunk(i - 1, 1)) { (acc, el) =>
+   *        el :: acc
+   *      }
+   *      .map(_.reverse)
+   *  )
+   *  .runCollect
+   * }}}
+   *
+   * The stream would emit the elements `List(1), List(4), List(1, 1)`.
+   * The [[foldWeightedDecomposeM]] allows the decompose function
+   * to return a `ZIO` value, and consequently it allows the transducer
+   * to fail.
+   */
+  def foldWeightedDecompose[I, O](
+    z: O
+  )(costFn: I => Long, max: Long, decompose: I => Chunk[I])(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    ZTransducer {
+      case class FoldWeightedState(started: Boolean, result: O, cost: Long)
+
+      val initial = FoldWeightedState(false, z, 0)
+
+      @tailrec def go(in: Chunk[I], os0: Chunk[O], state: FoldWeightedState): (Chunk[O], FoldWeightedState) =
+        in.headOption match {
+          case None => os0 -> state
+          case Some(i) =>
+            val total = state.cost + costFn(i)
+            if (total > max) {
+              val is = decompose(i)
+              if (is.isEmpty)
+                go(in.drop(1), os0 + f(state.result, i), initial)
+              else if (is.length == 1)
+                go(in.drop(1), os0 + f(state.result, is(0)), initial)
```

Comment:
We should avoid incrementally building chunks like this (at least until we have tree rebalancing in place). Let's try and build the entire output chunk at once?

Reply:
I've been looking at the rebalancing work from a distance. I don't really understand why we have `Singleton`, `Slice` and `Vector` chunks. Isn't locality of reference more important than space economy on tiny chunks?

Reply:
So Slice is there to make slicing operations (like take, drop, etc.) O(1) and zero-copy; similarly Vector is there because it's fine to keep as-is and avoid copying. Singleton is probably redundant because an array would occupy just an additional word in memory.